File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    void check(int idx,string digits,string temp,vector<string> &ans,vector<string>& dict){
        if(idx==digits.size()){ //possible answer
            ans.push_back(temp);
            return;
        }     
        string curr_string=dict[digits[idx]-'0'];  
        for(int i=0;i<curr_string.size();i++){       
            temp+=curr_string[i];
            check(idx+1,digits,temp,ans,dict);  
            temp.pop_back(); 
        }
    }
    vector<string> letterCombinations(string digits) {
        if(digits.size()==0) return {};
        vector<string> dict={"","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"}; 
        vector<string> ans; 
        string temp;
        int idx=0;  
        check(idx,digits,temp,ans,dict); //Time complexity o(n*4^n)
        return ans;
        
    }
};

File: 17-letter-combinations-of-a-phone-number/17-letter-combinations-of-a-phone-number.js
/**
 * @param {string} digits
 * @return {string[]}
 */

function check(idx,digits,temp,ans,dict)  {
    if(idx === digits.length){
        ans.push(temp)
        return;
    }
    let curr_str=dict[parseInt(digits[idx])];
    for(let i=0;i<curr_str.length;i++){
         temp+=curr_str[i];
         check(idx+1,digits,temp,ans,dict);
         temp = temp.substr(0, temp.length - 1);
    }
};

var letterCombinations = function(digits) {
    if(digits.length===0) return [];
    const dict =["","","abc","def","ghi","jkl","mno","pqrs","tuv","wxyz"];
    let ans=[];
    let temp="";
    check(0,digits,temp,ans,dict);
    return ans;
    
    
};